Refactor volume configurations

juanpabloguerra16 opened the pull request:
## Summary

This PR addresses storage compatibility issues with k8s distros that don't support hostpath or don't offer nvme `ReadWriteMany` storage classes and improves the chart's flexibility by implementing backward-compatible storage configuration options.

### Problem
The current storage configuration has several limitations:
- **GKE Incompatibility**: ReadWriteMany access mode is not supported with NVMe SSD ephemeral storage in GKE
- **Hardcoded Values**: Volume and PVC names are hardcoded, reducing flexibility
- **Shared PVC Anti-pattern**: All deployments share a single PVC (`gvs-shared-vol`), which creates resource contention and isn't a best practice for independent services

### Solution
Implemented backward-compatible storage configuration with two modes:

#### New Default (GKE-compatible)
- **Ephemeral Storage**: Uses `emptyDir` volumes with K8S local SSD ephemeral storage
- **Per-deployment Resources**: Each service has its own resource configuration and storage
- **No PV/PVC**: Eliminates persistent volume dependencies for cloud-native deployments

#### Legacy Support (Backward Compatible)
- **Persistent Storage**: Maintains PV/PVC support when `persistence.enabled: true`
- **Configurable Claims**: Supports both existing claims and automatic PVC creation
- **Flexible Configuration**: Allows customization of storage class, access modes, and sizes

### Key Changes
- Modified all deployment templates to conditionally handle volumes based on `persistence.enabled`
- Added per-deployment resource specifications with global fallback
- Created conditional PV/PVC templates that only deploy when persistence is enabled
- Updated `values.yaml` to default to ephemeral storage (`persistence.enabled: false`)

### Benefits
- ✅ **CSP K8S Compatible**: Works with ephemeral storage and local SSD
- ✅ **Backward Compatible**: Existing configurations continue to work
- ✅ **More Flexible**: Per-deployment resource allocation
- ✅ **Better Practices**: Eliminates shared storage anti-pattern
- ✅ **Cloud-Native**: Supports both persistent and ephemeral storage paradigms

Commit:
Refactor volume configurations in deployment templates to support conditional persistence

shivamerla commented on generative-virtual-screening-chart/templates/deployment_diffdock.yaml:
```diff
@@ -83,11 +83,15 @@ spec:
-            claimName: gvs-shared-vol
+            claimName: {{ .Values.persistence.existingClaim | default "gvs-shared-vol" }}
+          {{- else }}
+          emptyDir: {}
```
I don't think emptyDir is a good choice for model storage. These models are large, and using node-local storage increases the risk of eviction on shared systems. Better to use CSI-backed ephemeral volumes which support proper storage isolation and resource tracking. For GKE, something like [this](https://cloud.google.com/kubernetes-engine/docs/how-to/generic-ephemeral-volumes) would be more appropriate.

juanpabloguerra16 replied:
Thanks Shiva, this is helpful. I agree but since this blueprint requires NVME SSD because there's a POD that requires 1.5 TiB (msa) PD would be too slow. The ephemeral volume that you linked requires to set emptyDir on GKE. FWIW the primary goal of this PR is to make the storage configurable as atm it is not. Maybe we can inject emptyDir as part of `values.yaml` instead of hardcoding it

juanpabloguerra16 replied:
<img width="502" height="314" alt="image" src="https://github.com/user-attachments/assets/116d6870-5dc0-45b9-8f96-ad1b7590f6bb" />

juanpabloguerra16 replied:
https://github.com/NVIDIA-BioNeMo-blueprints/generative-virtual-screening/pull/9/commits/a427adb5c801bb648e584493d997535f19d5a9a1

Commit:
refactor: replace binary persistence toggle with explicit volume type configuration

- Add configurable volume types: pvc, hostPath, emptyDir with clear use cases
- Delete redundant PV template that wrapped hostPath in unnecessary abstraction
- Eliminate ambiguous fallback logic for consistent, explicit configuration

edawson reviewed (approved):
I think this looks good, though I'm going to trust you tested this locally @juanpabloguerra16 as I have no way to test.

juanpabloguerra16 committed:
add volume mount name to msa
